### ISSUE TITLE
New Language Support: Powershell

### ISSUE THREAD

**Issue by JustinGrote** [pulumi/pulumi#3507]
Now that the .NET provider is available, I have created a proof of concept to enable Powershell to be used with Pulumi.

https://github.com/JustinGrote/PulumiPowershellTest
![Pulumi Preview](https://raw.githubusercontent.com/JustinGrote/PulumiPowershellTest/master/images/PulumiPSDemo.gif)

I am creating this issue as a discussion for the work to make this a more viable production reality. I am not a C# developer and am very bad at it, I had to look up pretty much everything to get this POC to work :)

Likely Required Work Items:

1. Create a robust pulumi powershell host that derives from the [PSHost](https://docs.microsoft.com/en-us/dotnet/api/System.Management.Automation.Host.PSHost?view=pscore-6.2.0) class that runs the script asynchronously and reads the Powershell streams to update the status notes from verbose/warning/error/etc messages in real time.
2. Process Powershell Output Stream into Outputs intelligently
3. Run within existing local scope, or pass items so that the DLLs don't have to be instantiated twice.
4. Add another Pulumi CLI runtime`powershell` that would use a precompiled powershell host to run the script, so that dotnet restore doesn't have to happen every time and will speed up the execution process. If a user needs more .NET types, they can add them at runtime with `Add-Type -Path /path/to/my.dll`. This would be defined by setting the pulumi.yaml runtime to `powershell`

Good places where a lot of similar code required would be the AWS Lambda Powershell Core runner and the Azure Functions Powershell Worker, they solve many of these problems already.

POC Tested on Windows 10 and Ubuntu 18.04, works on both. Haven't tested Mac.

@mikhailshilkov FYI

**Comment by bergmeister** [comment]
This is a great start but as an end user, having something more DSC or Terraform like would be awesome in terms of syntax

**Comment by JustinGrote** [comment]
Why wouldn't you just use terraform if you want a declarative style syntax?

**Comment by bergmeister** [comment]
Terraform has the disadvantage that the language has limitations on what one can do in terms of logic. In PowerShell I can  write a function that can do literally anything

**Comment by mikhailshilkov** [comment]
@bergmeister Note that Pulumi (including Pulumi PowerShell) is still based on the idea of desired state configuration. Whenever you define, say, a resource group - it's not immediately translated to the API calls - it builds up the resource graph. Then, the CLI compares that target graph to the current state and issues the diff. That's the important difference to "normal" PowerShell scripts.

Syntax suggestions are very welcome: now is the best time to add your thoughts. Thank you!

**Comment by bergmeister** [comment]
Awesome, I didn't know that, just by looking it I though it was imperative and not declarative, thanks for clarifying that.
My suggestion as a later improvement would be to support 'proper' nodes of the DSC like syntax where scriptblocks are used instead of calling the constructor via `::new` and having to wire up the arguments, which is a bit tedious. It could look like this, which is neater to my, IMGO
```pwsh
ResourceGroup {
    Location = Get-LocationFromComplicatedFunction
    Name = $MyName
}
```

**Comment by mikhailshilkov** [comment]
@bergmeister Yeah, that's a bit of a counter-intuitive point, I try to explain it every time I introduce Pulumi to somebody.

Your example does look much cleaner! To be honest, I'm not fluent with PowerShell so I have no idea how to implement such a construct. Is it a lot of work to manually prototype it for a Resource Group, and maybe a Storage Account? Then we could take a look at generating it for all the resources.

Thanks a ton for contributing!

**Comment by bergmeister** [comment]
The construct that I am describing is a `scriptblock`, one sees this a lot as it allows for defining 'things' in an elegant way that make it look like it is descriptive, for example in PowerShell's test framework that looks like:
```pwsh
Describe 'My Test Suite name {
   Beforeach {
       # Perform action before each test
   }
    It 'My test name 1' {
        Get-AnswerToTheMeaningOfLifeUniverseAndEverything | Should -Be 42
    }
}

**Comment by JustinGrote** [comment]
@bergmeister, what you're suggesting is what terraform already does though. It's a DSL, so why reinvent the wheel? The point of Pulumi is to use native code to get away from the static limitations of a DSL, but everything I show in this example is both declarative (if you modify the code and change the name of the resource groups, it will destroy the old ones before creating the new ones. It also shows you a preview plan, I just skipped the approval in the gif for brevity with the -y switch) and automatic dependency handling is used as well. 

Using types is a bit of a DSL like syntax anyways, I'll post some examples. 

There's many ways to skin a cat in powershell, so once this basic runtime is sorted and the objects are available, you can build on both functions and a DSL in a powershell module if you want. 

It would be theoretically possible to dynamically generate a DSL by iterating through the imported types and creating functions, but that would be more of a "2.0" type feature I think.

**Comment by bergmeister** [comment]
The big difference to terraform is that terraform has to create its own small language that has limitations.
PowerShell is a language that is already mature with all the bells and whistles that are needed and all one would do here is prettify the usage of calling a constructor and populating the object.

**Comment by JustinGrote** [comment]
@bergmeister I agree but the DSL will have its own challenges. How do you do a stable generation of the DSL code? If it is hand-written then it has to be constantly updated. If it is procedurally generated it may not be able to cope with all the eccentricities of the types that come through.

Let me throw up an update to the proof of concept that explains what you're talking about so @mikhailshilkov has an idea. I also welcome as many easy avenues as possible, but for the scope of this issue I think it should be focused on getting it "out the door" in a functional and stable way as soon as possible since unlimited resources are probably not available at Pulumi :).

That said, a DSL would be a great bolt-on, I'll show a working example that @mikhailshilkov asked for in a bit.

**Comment by mikhailshilkov** [comment]
Basically, all Pulumi providers' code is generated, so code generation is not a huge problem in itself. If we nail the shape of objects, we can add PowerShell-specific codegen on top of C# codegen. I'm toying with the same idea for F#, for example.

Of course, the fewer variants, the easier to support, so an extra layer must give enough improvement to the user experience to warrant its existence. Also think of code samples, docs, etc.

**Comment by JustinGrote** [comment]
@bergmeister  @mikhailshilkov  I've put up 6 different styles (including a DSL) into a new branch. 
https://github.com/JustinGrote/PulumiPowershellTest/blob/multipleSyntaxExamples/pulumi.ps1

They all work:
![image](https://user-images.githubusercontent.com/15258962/68879401-34f13f80-06be-11ea-88f3-a884e8d5cd2e.png)

**Comment by JustinGrote** [comment]
So, that being said, a DSL or functions could be created either at the powershell level or at the c# level by generating a binary module. I think powershell-level would make most sense since it can be dynamically instituted on load from the available providers whereas dotnet would have to be recompiled every run and I think the goal should be to have a precompiled powershell host to run the scripts, rather than through my "dotnet build" shim.

**Comment by bergmeister** [comment]
Yes, I agree that you guys have to ship a MVP first to deliver value iteratively. I just wanted to provide an idea what a more polished version would look like, I agree this is more v2 or something the community might do (which I doubt though)

**Comment by JustinGrote** [comment]
Here's an example of a "close" DSL format that doesn't require any extra implementation (this deploys an azure function):

```powershell
    $Name = 'myTestFunctionApp'
    $resourceGroup = [ResourceGroup]::new($Name)

    $functionApp = [FunctionApp]::new($Name, @{
        ResourceGroupName = $resourceGroup.Name
        StorageConnectionString = [Account]::new($Name,@{
            ResourceGroupName      = $resourceGroup.Name
            AccountReplicationType = 'LRS'
            AccountTier            = "Standard"
        }).PrimaryConnectionString
        AppServicePlanId = [Plan]::new($Name,@{
            ResourceGroupName = $resourceGroup.Name
            Kind = 'FunctionApp'
            Sku = [PlanSkuArgs]@{
                Tier = 'Dynamic'
                Size = 'Y1'
            }
            
        }).Id
    })
```

**Comment by manoj2994** [comment]
Pulumi for native powershell cmdlet support will be better than a dsl .in my point of view ,if it not in dsl we don't need to go back and check with the new syntaxes that would be little bit pain.if iam not wron pulumi for native programming language support so we can use easily without any prior dsl knowledge like terraform .so i would say pulumi supports for native powershell will be great

**Comment by JustinGrote** [comment]
@manoj2994 The idea is there is more than one way to skin a cat in Powershell. First make objects available, then abstractions (functions, classes, DSL) can be built on top of that.

The DSL doesn't have to be a static DSL, it would work like Pester works. For instance, this totally works per example 6 above

```powershell
1..5 | foreach {
  StorageAccount "dslstorage$PSItem" @{
        ResourceGroup = (ResourceGroup "DSLResourceGroup$PSItem")
  } > $null
}
```

Will make 5 storage accounts. The "DSL" in powershell is just tersely named functions with a specific parameter format, they don't mean we have to go to a "static" YAML or JSON layout like terraform/ARM

**Comment by mikhailshilkov** [comment]
@JustinGrote If we make a standard host, it would only reference the main `Pulumi` NuGet package. What would be the idiomatic way to reference other NuGet packages like `Pulumi.Azure` and load those DLLs from the end-user program?

**Comment by JustinGrote** [comment]
You could either add them on startup with Add-Type, or maybe make pseudo Powershell Modules that all they do is load the assemblies, for instance Import-Module -Name Pulumi.Azure. There's lots of ways to go about that.

**Comment by mikhailshilkov** [comment]
@JustinGrote To use `Add-Type` one first needs to restore the NuGet package and have a path to reference, right?

**Comment by JustinGrote** [comment]
Correct, typically for PowerShell modules you embed the dll assemblies and
put the add-type in the psm1. You could also have a module with cmdlets to
dynamically import.

For now I would just focus on getting the basic powershell host implemented
and verify all the gRPC communication works as intended.

**Comment by JustinGrote** [comment]
@mikhailshilkov @leezen I notice this was dropped from all milestones. Is this no longer a priority?

**Comment by leezen** [comment]
We didn't have a chance to work on this the last few months. We're examining our roadmap priorities for the rest of Q2 to see how this might slot in.

**Comment by leezen** [comment]
Just an update here: with the current plans, we don't expect to work on PowerShell support this quarter. We'll re-examine the priorities for this issue next quarter (starting July 2020).

Separately, we should do a better job documenting how to get any existing support working and I've created https://github.com/pulumi/docs/issues/3213 to track that.

**Comment by 4c74356b41** [comment]
Hey, its almost July, can you update us, please? @leezen

**Comment by leezen** [comment]
I'll have a concrete update next week when we discuss Q3 roadmap.

**Comment by leezen** [comment]
This won't be prioritized in Q3 by the team. Definitely open to community contributions here!

**Comment by JustinGrote** [comment]
On the same day Amazon CDK announced support for Terraform (and azure via the provider), guess I'll invest over there then...

**Comment by leezen** [comment]
@JustinGrote Sorry this was disappointing news to you. I should clarify in case it wasn't clear that we still very much want to support Powershell. While we currently don't plan to work on this issue, we consistently re-examine our backlog and priorities both on a bi-weekly basis as well as quarterly.

**Comment by rdbartram** [comment]
Hi guys, I've wanted to start on this for a while but couldn't get the time. I have a proof of concept locally of grpc client running natively in PowerShell. I will see if I can't get some basic communication going on between the pulumi executable and pwsh...Sorry for anyone wanting WinPS support but i'm focusing solely on PSCore.

**Comment by JustinGrote** [comment]
@rdbartram you may want to look at my proof of concept, the .NET objects work natively, you don't have to roll your own GRPC, a Powershell Host basically just has to be created to make it more proper:
https://github.com/JustinGrote/PulumiPowershellTest

Several examples of it in use:
https://github.com/JustinGrote/PulumiPowershellTest/blob/multipleSyntaxExamples/pulumi.ps1

**Comment by rdbartram** [comment]
Ok, I'm pretty excited guys. I have a PoC locally now which natively starts pwsh from go. It loads the dlls built from the dotnet project which i think is fine and it creates the expected resources...will test further and hopefully start a draft PR in the coming days

**Comment by rdbartram** [comment]
![pulumi_pwsh](https://user-images.githubusercontent.com/16256546/87839551-326d6800-c893-11ea-95ae-76149faaaba7.gif)

**Comment by rdbartram** [comment]
Has anyone seen an issue where the resources that are being passed to pulumi somehow are not always getting read properly...i've got a weird issue where i can run a preview multiple times with something as basic as resource groups and it sometimes says it has to delete...then nothing, then create...Seemingly only happening in this pwsh implementation as when I went back to @JustinGrote's example, I wasn't getting the issue. I'm wondering if I'm missing a trick or if something is fundamentally wrong with the pwsh implementation. I would expect it to work or not...not sometimes

**Comment by mikhailshilkov** [comment]
@rdbartram that definitely sounds wrong and is probably caused by some implementation issues.

Could you post a sample program, stack state at the moment of preview (state is in `pulumi stack export`), and the preview results?

**Comment by madsaune** [comment]
Hi, has there been any updates related to this?

**Comment by rdbartram** [comment]
sorry this fell off my radar as I got busy with other projects and the c word...I'll see if I can't dig out this example again and at least post it to github so others can take a look at it. The status is as I last posted here...I have an implementation with pwsh which works kind of...I can natively create the objects and it all shows up gravy...there was just a wierd bug with it sporadically not seeing objects getting created and so marking them for deletion

**Comment by JustinGrote** [comment]
@rdbartram you got a repo anywhere? I couldn't find an appropriate branch in your pulumi fork :)

**Comment by cveld** [comment]
@rdbartram has published their code to https://github.com/worxspace/pspulumi
They recently gave a conference talk which got published on YouTube: https://www.youtube.com/watch?v=YAQLvNjJKOs

**Comment by Frassle** [comment]
> They recently gave a conference talk which got published on YouTube

Interesting that they we're looking at full language provider and then changed to just emit yaml to the pulumi-yaml language provider.

We've got _a lot_ of ideas to make it easier for the community to do full language providers, so that might be worth revisiting at some point.

**Comment by fsdrw08** [comment]
any update of this request? I look forward to this.

**Comment by DannyArr** [comment]
Looking forward to this as well.

**Comment by aescolastico** [comment]
bump.
